common/src/main/kotlin/io/homeassistant/companion/android/common/data/websocket/impl/WebSocketCoreImpl.kt

```diff
@@ -0,0 +1,543 @@
+
+    override suspend fun sendMessage(request: WebSocketRequest): SocketResponse? {
+        return if (connect()) {
+            withTimeoutOrNull(request.timeout) {
+                try {
+                    suspendCancellableCoroutine { cont ->
+                        // Lock on the connection so that we fully send before allowing another send.
+                        // This should prevent out of order errors.
+                        connection?.let {
+                            synchronized(it) {
+                                val requestId = id.getAndIncrement()
+                                val outbound = request.message.plus("id" to requestId)
+                                Timber.d("Sending message $requestId: $outbound")
+                                activeMessages[requestId] = request.apply {
+                                    onResponse = cont
+                                }
+                                connection?.send(webSocketMapper.writeValueAsString(outbound))
+                                Timber.d("Message number $requestId sent")
+                            }
+                        }
+                    }
+                } catch (e: Exception) {
+                    Timber.e(e, "Exception while sending message")
+                    null
+                }
+            }
+        } else {
+            Timber.w("Unable to send message, not connected: $request")
+            null
+        }
+    }
+
+    override suspend fun sendBytes(data: ByteArray): Boolean? {
+        return if (connect()) {
+            withTimeoutOrNull(30.seconds) {
+                try {
+                    connection?.let {
+                        synchronized(it) {
+                            it.send(data.toByteString())
+                        }
+                    }
+                } catch (e: Exception) {
+                    Timber.e(e, "Exception while sending bytes")
+                    null
+                }
+            }
+        } else {
+            Timber.w("Unable to send bytes, not connected")
+            null
+        }
+    }
+
+    /**
+     * Start a subscription for events on the websocket connection and get a Flow for listening to
+     * new messages. When there are no more listeners, the subscription will automatically be cancelled
+     * using `unsubscribe_events`. If the subscription already exists, the existing Flow is returned.
+     *
+     * @param type value for the `type` key in the subscription message, for example `subscribe_events`
+     * @param data a key/value map of additional data to be included in the subscription message, for
+     *             example the `event_type` + value when subscribing with `subscribe_events`
+     * @param timeout timeout until the subscription is ended after the flow is no longer collected
+     * @return a Flow that will emit messages delivered to this subscription, or `null` if an error
+     *         occurred
```

Comment:
Shouldn't this (also) be on the interface, so users can actually use the documentation? You're describing parameters here, not just internals.

Reply:
(Technically I'm describing it here as the original author, but you moved it 😉)

Reply:
I'm moving it to the interface. I plan to rework the whole documentation of the file in a future PR.